### PR TITLE
schedule/labeler: index keyspace rules incrementally

### LINKED ISSUE #11030: [scalability] Index keyspace region-label rules incrementally
### Problem

Adding or deleting one canonical keyspace region-label rule rebuilds and sorts the complete generic range index while holding the RegionLabeler exclusive lock. Scheduling Service repeats that rebuild after its etcd watch event. At large keyspace counts this makes single-keyspace operations and restart readiness depend on O(K log K) work and a high transient memory peak.

### Expected behavior

- A single keyspace-rule add/delete does not traverse or sort all existing rules.
- RawKV-disabled and RawKV-enabled label semantics remain unchanged.
- Startup indexing has bounded memory and a documented readiness target.
- Scale benchmarks cover startup and lock-held single-rule updates at 100k and 1M rules.

This is the upstream counterpart of https://github.com/tidbcloud/pd-cse/issues/560.

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

Issue Number: Close #11030

Adding or deleting one canonical keyspace region-label rule rebuilt and sorted
the complete generic range index under the labeler lock. The cost and transient
memory therefore grew with every existing keyspace and was repeated by the
Scheduling Service watcher.

This is the upstream counterpart of tidbcloud/pd-cse#560.

### What is changed and how does it work?

- Recognize only the exact RawKV/Txn keyspace-rule shape generated by
  pkg/keyspace; all other rules retain the generic range-list behavior.
- Store canonical rules in sparse fixed-size slots with presence bitsets and
  derive their split boundaries on demand.
- Update one canonical rule in bounded time without traversing or sorting all
  rules, including Plan and watcher paths.
- Merge canonical and generic split points during lookup while preserving
  RawKV-disabled, RawKV-enabled, generic-rule, and priority semantics.
- Keep lookups allocation-free and avoid duplicate generic range-list searches.
- Add edge-case, collision-atomicity, and cross-boundary tests plus 100k/1M
  startup, lock-held update, and lookup benchmarks.

```commit-message
schedule/labeler: index keyspace rules incrementally
```

### Check List

Tests

- [x] Unit test
- [x] Integration test
- [x] Manual scale benchmark

Self-checks:

- make gotest GOTEST_ARGS='./pkg/schedule/labeler ./pkg/schedule/rangelist ./pkg/mcs/scheduling/server/rule -count=1'
- go test -race on the changed labeler/rangelist tests
- make static, followed by final targeted golangci-lint
- 100k/1M benchmarks: startup 144 ms / 1.70 s, peak RSS 89 MB / 668 MB;
  lock-held update 156 ns / 163 ns and lookup 100 ns / 104 ns, all with 0 B
  and 0 allocations per operation

Code changes

- No configuration, HTTP API, or persistent-data format changes.

Side effects

- Adds a specialized in-memory index for deterministic keyspace rules; generic
  user rules continue to use the existing range list.

Related changes

- Companion fix: tidbcloud/pd-cse#560

### Release note

```release-note
Improve region-label scalability for large keyspace counts by indexing deterministic keyspace rules incrementally.
```

## Summary by CodeRabbit

- **New Features**
  - Added deterministic keyspace-based region labeling with consistent results for both raw and transactional modes.
  - Enhanced split-key discovery by merging keyspace and generic range outputs and resolving boundaries strictly within open intervals.

- **Bug Fixes**
  - Keyspace rule updates/deletions now update only the relevant deterministic label slots, preventing unintended label/range-list side effects.
  - Improved correctness when combining generic rules with keyspace boundaries, including edge-range handling.

- **Performance**
  - Introduced incremental, sparse indexing for generic and keyspace rules to reduce rebuild overhead.

- **Tests**
  - Added comprehensive unit tests and benchmarks covering indexing, boundary behavior, legacy compatibility, incremental updates, and sparse reuse.

### CODE REVIEW

File: pkg/schedule/rangelist/range_list.go [hunk 1]
```diff
@@ -74,3 +74,24 @@ func (l List) GetSplitKeys(start, end []byte) [][]byte {
...
+	if i < len(l.segments) && (len(end) == 0 || bytes.Compare(l.segments[i].startKey, end) < 0) {
+		return nil, false
+	}
+	if i == 0 {
+		return nil, true
+	}
+	return l.segments[i-1].data, true
+}
+
+// HasSplitKey reports whether there is a split point in (start, end).
```
Comment (lhy1024) [inline pkg/schedule/rangelist/range_list.go]: **P3 - Remove the unused `HasSplitKey` API**

`rangelist.List.HasSplitKey` is only referenced by tests. The production path uses `GetDataByRange` directly, which already obtains both the range data and split status in one lookup. Keeping `HasSplitKey` adds a new exported API without a consumer. Please remove it and its test coverage unless there is an intended production caller.

File: pkg/schedule/labeler/keyspace_index.go [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+		return
+	}
+	slot := int(id) & keyspaceChunkMask
+	chunk.rules[slot] = nil
+	chunk.bits[slot>>6] &^= uint64(1) << (id & 63)
+	chunk.count--
+	if chunk.count == 0 {
+		s.chunks[chunkID] = nil
+		for len(s.chunks) > 0 && s.chunks[len(s.chunks)-1] == nil {
+			s.chunks = s.chunks[:len(s.chunks)-1]
```
Comment (lhy1024) [inline pkg/schedule/labeler/keyspace_index.go:111]: **P2: canonical-rule replacement still regresses for single-slot and high-ID chunks.**

When the current rule is the only entry in its chunk, `clear` drops the chunk and the following `set` allocates a new roughly 8 KiB `keyspaceRuleChunk`. For a high ID, this loop also trims every trailing nil directory entry, then `set` grows the directory again. This is reachable rather than theoretical: NextGen uses `SystemKeyspaceID = MaxValidKeyspaceID - 1`.

I verified this against the current PR head with a temporary targeted test and benchmark:
- dense chunk: `99 ns/op, 0 B/op, 0 allocs/op`
- single low ID: `4552 ns/op, 9473 B/op, 1 alloc/op`
- single max ID: `14113 ns/op, 9472 B/op, 1 alloc/op`

This contradicts the claimed allocation-free lock-held update path and leaves a sparse high-ID regression uncovered. Please avoid destroying/reallocating the sole chunk during replacement and avoid trimming then regrowing the high-ID directory; add a single-slot/high-ID benchmark or test.

Comment (rleungx) [reply pkg/schedule/labeler/keyspace_index.go:111]: Fixed in a50a69324. Canonical replacements now update owned slots in place, preserving the existing chunk and directory for both single-slot and max-ID rules. Added a zero-allocation reuse test for low and max IDs. Verification: full labeler/rangelist package tests passed; single-rule-update/100000 is 79.40 ns/op, 0 B/op, 0 allocs/op.

**Review by lhy1024 — APPROVED** [review]

**Comment by rleungx** [comment]
@JmPotato PTAL

File: pkg/schedule/labeler/keyspace_index_test.go [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+		for range b.N {
+			if keys := index.GetSplitKeys(nil, nil); len(keys) != 2 {
+				b.Fatalf("expected 2 split keys, got %d", len(keys))
+			}
+		}
+	})
+}
+
+func BenchmarkRegionLabelerKeyspaceIndex(b *testing.B) {
+	for _, count := range []int{100_000, 1_000_000} {
```
Comment (JmPotato) [inline pkg/schedule/labeler/keyspace_index_test.go:395]: Issue #11030 requires a documented readiness target. These benchmarks measure startup time, but the repository does not define an acceptable limit; the numbers in the PR description are only one observation. Please document the target and its scope so future regressions have a clear acceptance criterion.

**Review by JmPotato — APPROVED** [review]
